x/mongo/driver/topology/CMAP_spec_test.go

```diff
@@ -112,26 +150,66 @@ func runCMAPTest(t *testing.T, testFileName string) {
 		backgroundThreadErrors: make(chan error, 100),
 	}
 
-	l, err := net.Listen("tcp", "localhost:0")
-	testHelpers.RequireNil(t, err, "unable to create listener: %v", err)
-
-	s, err := NewServer(address.Address(l.Addr().String()),
-		primitive.NewObjectID(),
-		WithMaxConnections(func(u uint64) uint64 {
+	sOpts := []ServerOption{
+		WithMaxConnections(func(uint64) uint64 {
 			return uint64(test.PoolOptions.MaxPoolSize)
 		}),
-		WithMinConnections(func(u uint64) uint64 {
+		WithMinConnections(func(uint64) uint64 {
 			return uint64(test.PoolOptions.MinPoolSize)
 		}),
-		WithConnectionPoolMaxIdleTime(func(duration time.Duration) time.Duration {
+		WithMaxConnecting(func(uint64) uint64 {
+			return uint64(test.PoolOptions.MaxConnecting)
+		}),
+		WithConnectionPoolMaxIdleTime(func(time.Duration) time.Duration {
 			return time.Duration(test.PoolOptions.MaxIdleTimeMS) * time.Millisecond
 		}),
-		WithConnectionPoolMaintainInterval(func(duration time.Duration) time.Duration {
+		WithConnectionPoolMaintainInterval(func(time.Duration) time.Duration {
 			return time.Duration(test.PoolOptions.BackgroundThreadIntervalMS) * time.Millisecond
 		}),
-		WithConnectionPoolMonitor(func(monitor *event.PoolMonitor) *event.PoolMonitor {
-			return &event.PoolMonitor{func(event *event.PoolEvent) { testInfo.originalEventChan <- event }}
+		WithConnectionPoolMonitor(func(*event.PoolMonitor) *event.PoolMonitor {
+			return &event.PoolMonitor{
+				Event: func(evt *event.PoolEvent) { testInfo.originalEventChan <- evt },
+			}
+		}),
+	}
+
+	// If there's a failpoint configured in the test, use a dialer that returns connections that
+	// mock the configured failpoint. If "blockConnection" is true and "blockTimeMS" is specified,
+	// use a mock connection that delays reads by the configured amount. If "closeConnection" is
+	// true, close the connection so it always returns an error on read and write.
+	if test.FailPoint != nil {
```

Comment:
This looks great; clever workaround to create immediately-closed connections and delayed connections for closing and blocking failpoints respectively.